src/core/infer_request.h

```diff
@@ -337,6 +346,15 @@ class InferenceRequest {
     return Status::Success;
   }
 
+  // Add a callback to be invoked on destruction. Multile callbacks
+  // can be added by calling this function in order, and they will be
+  // invoked in reversed order.
+  Status AddInternalCallback(std::function<void()>&& callback)
```

Comment:
AddReleaseCallback.
These should be invoked when the request is released by Triton. The client can reuse the request or even just hold onto it for a long time and we don't want to defer these callbacks until then.